pkg/nmi/server/server.go

```diff
@@ -244,7 +271,7 @@ func (s *Server) getMICToken(logger *log.Entry, rqClientID, rqResource string) (
 		// TODO: return the right status code based on the error we got from adal.
 		return nil, http.StatusForbidden, err
 	}
-	response, err := json.Marshal(*token)
+	response, err := json.Marshal(newMSIResponse(*token))
```

Comment:
Is there a way to check the sdk version from which the call was originating and then return back either the new format or the new format. That would make it work for both versions. What do you say ?

Reply:
That's an interesting thought. @aramase had a similar question just below. I'll answer on this thread.

It seems like the most important thing is for the NMI response body to be as nearly identical to the underlying IMDS endpoint as possible. Do we know if the original response body returns unquoted numbers based on the `api-version` on the request? Let's see...

Reply:
It does look like the api specification for even the latest version declares those fields as `string` in the [IdentityTokenResponse](https://github.com/Azure/azure-rest-api-specs/blob/8f2e7ec/specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2019-06-01/imds.json#L868).

I suspect the change to `type json.Number string` in the `adal.Token` was more for convenience in end-user code, and they didn't anticipate that type would be used to round-trip Marshal the data back into a response wire-form.

Reply:
Confirmed on a VM also - looks like those are actually string fields even on the newest api-version.

```
curl 'http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-10-01&resource=https://management.core.windows.net/' --header 'Metadata: true' -D -

HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Server: Microsoft-IIS/10.0
Date: Wed, 24 Jul 2019 22:07:19 GMT
Content-Length: 1683

{"access_token":"eyJ0--(redacted)--b71c","expires_in":"28800","expires_on":"1564028884","ext_expires_in":"28800","not_before":"1563999784","resource":"https://management.core.windows.net/","token_type":"Bearer"}
```

Reply:
@lodejard - I don't think there was a change in IMDS, but a change in the adal sdk. The core issues from https://github.com/Azure/go-autorest/pull/326 talks about the reconciling the difference between  AAD and ADFS. If we can make NMI response corresponding to the version which is used by application, then we will be able to get it working with old and new adal versions. What do you say ?

Reply:
After taking a look at the api contracts, [they're here](https://github.com/Azure/azure-rest-api-specs/blob/8f2e7ec/specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2019-06-01/imds.json#L868), and curling the actual endpoint on a VM with the latest `api-version` (comment above) --- the metadata endpoint does always return strings and never numbers for all of those fields.

I would say it's more correct to always return strings, just like the metadata service does. In that way this PR should work on old and new adal client versions because the NMI response will be identical to what adal is already unmarshalling from the metadata endpoint when NMI is not installed.

Reply:
Here's another way to look at it...

Unmarshal to json.Number works with both wire-forms, so for the client-side like adal it's the best choice for data type.
```
{"example": "42"} -> json.Number=42
{"example": 42}   -> json.Number=42
```

But Marshal from json.Number breaks any client that unmarshals with string, so for the server-side like NMI it causes compat problems
```
json.Number=42 -> {"example": 42} -> json.Number=42
json.Number=42 -> {"example": 42} -> string=Error
```

So even the basic golang example in the ms docs couldn't unmarshal an aad-pod-identity response if it has unquoted numbers.
https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-go

Reply:
@lodejard Thank you for trying it out. I also went through the code in the sdk and what you are saying makes total sense.